File: src/SilksongLauncher.Launcher/app/src/main/kotlin/dev/silksong/launcher/skins/core/SkinRotationTransactionModels.kt
package dev.silksong.launcher.skins.core

import java.util.Collections
import dev.silksong.launcher.skins.registry.*

/** Origin is retained unchanged, including both authorities and exact issued readiness.
 * Qualified durable recovery retires it into one bounded historical audit. */
data class RotationTransactionOperation(val origin: RotationModeState, val readiness: RotationReadyIntent?,
    val transaction: TransactionState, val reboundBlocked: Boolean = false)
data class RotationTransactionState(val mode: RotationModeState, val operation: RotationTransactionOperation? = null,
    val readinessReboundBlocked: Boolean = false, val lastRecovery: RotationRecoveryRecord? = null)
/** Exact current authority supplied by the trusted caller, not discovered by this reducer. */
data class RotationRecoveryAuthority(val hero: HeroBindingToken, val skin: SkinBindingToken,
    val liveProof: HeroVerifiedVisual? = null)
/** Trusted serialized outcome attestation: ALL old execution and late callbacks are
 * irrevocably fenced, and this exact parent was authoritative at the closure CAS.
 * The reducer verifies correlation/chain only; constructing this value is not runtime proof.
 * Optional lateFailureReceipt is ONE ARM-to-failure hop, only if no failure receipt was
 * stored and both original failure codes are known. It never replaces stored authority. */
data class RotationOutcomeFence(val correlation: TransactionCorrelation, val originalHero: HeroBindingToken,
    val authoritativeParent: VerifiedRegistryHead, val lateFailureReceipt: RegistryCommitReceipt? = null)
sealed interface RotationRecoveryEvidence {
    val authority: RotationRecoveryAuthority
    /** Resolution of the exact already-issued pending closure, never an outcome selector. */
    data class IssuedClosureOutcome(val operation: RotationTransactionOperation, val fence: RotationOutcomeFence,
        val receipt: RegistryCommitReceipt, val head: VerifiedRegistryHead,
        override val authority: RotationRecoveryAuthority) : RotationRecoveryEvidence
    /** Trusted completed original-binding prior restoration, serialized with the fence and
     * exact authoritative parent. A possibly committed target cannot be superseded by a
     * visual observation: the verifier must establish this whole serialized outcome. */
    data class PriorRestoredAndFenced(val operation: RotationTransactionOperation, val fence: RotationOutcomeFence,
        val restoration: HeroVerifiedVisual, val receipt: RegistryCommitReceipt, val head: VerifiedRegistryHead,
        override val authority: RotationRecoveryAuthority) : RotationRecoveryEvidence
    data class VisualClosure(val operation: RotationTransactionOperation,
        val completion: TransactionEvent.CompletionCommitted, override val authority: RotationRecoveryAuthority) : RotationRecoveryEvidence
    /** All old execution deliveries irrevocably fenced; visual Apply never ran.
     * Prepare/ARM may have occurred. A head observation alone is insufficient. */
    data class VisualFencedNotExecuted(val operation: RotationTransactionOperation, val head: VerifiedRegistryHead,
        val receipt: RegistryCommitReceipt? = null, override val authority: RotationRecoveryAuthority) : RotationRecoveryEvidence
    data class ModeClosure(val operation: ModeOperation, val completion: RotationModeEvent.ModeCommitted,
        override val authority: RotationRecoveryAuthority) : RotationRecoveryEvidence
    /** Exact intent never consumed/executed; all its execution deliveries fenced. */
    data class ReadinessFencedNotExecuted(val intent: RotationReadyIntent, val head: VerifiedRegistryHead,
        override val authority: RotationRecoveryAuthority) : RotationRecoveryEvidence
}
/** One bounded audit snapshot: before.lastRecovery must be null. No resource release authority. */
data class RotationRecoveryRecord(val before: RotationTransactionState, val evidence: RotationRecoveryEvidence)
sealed interface RotationTransactionEvent {
    data class Recover(val evidence: RotationRecoveryEvidence) : RotationTransactionEvent
    data class Mode(val event: RotationModeEvent) : RotationTransactionEvent
    data class ConsumeReady(val intent: RotationReadyIntent) : RotationTransactionEvent
    data class Transaction(val hero: HeroBindingToken, val event: TransactionEvent) : RotationTransactionEvent
}
sealed interface RotationTransactionCommand {
    data class Mode(val command: RotationModeCommand) : RotationTransactionCommand
    /** Caller must check the original hero as well as the inner command's skin correlation. */
    data class Transaction(val hero: HeroBindingToken, val command: SkinCommand) : RotationTransactionCommand
}
class RotationTransactionDecision(val state: RotationTransactionState, val diagnosis: String,
    commands: List<RotationTransactionCommand> = emptyList()) {
    val commands: List<RotationTransactionCommand> = Collections.unmodifiableList(commands.toList())
}

File: tools/hollow-knight-patches/src/skins/core/SkinRotationTransactionModels.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace DualSouls.Skins.HollowKnight.Core
{
    // Origin retains exact hero+skin and issued readiness. Qualified durable recovery
    // retires it into one bounded historical audit.
    public sealed record RotationTransactionOperation(RotationModeState Origin, RotationReadyIntent Readiness,
        TransactionState Transaction, bool ReboundBlocked = false);
    public sealed record RotationTransactionState(RotationModeState Mode, RotationTransactionOperation Operation = null,
        bool ReadinessReboundBlocked = false, RotationRecoveryRecord LastRecovery = null);
    // Trusted current authority, not discovered by the reducer.
    public sealed record RotationRecoveryAuthority(HeroBindingToken Hero, SkinBindingToken Skin, HeroVerifiedVisual LiveProof = null);
    // Trusted serialized outcome attestation: ALL old execution and late callbacks are
    // irrevocably fenced; this exact parent was authoritative at the closure CAS. The core
    // validates value correlation/chain, not runtime truth or construction authority.
    // Optional LateFailureReceipt is ONE ARM-to-failure hop when both failure codes are
    // known and no receipt was stored. It can never replace stored failure authority.
    public sealed record RotationOutcomeFence(TransactionCorrelation Correlation, HeroBindingToken OriginalHero,
        VerifiedRegistryHead AuthoritativeParent, RegistryCommitReceipt LateFailureReceipt = null);
    public abstract record RotationRecoveryEvidence(RotationRecoveryAuthority Authority)
    {
        // Exact already-issued stored pending closure; not a caller-selected target.
        public sealed record IssuedClosureOutcome(RotationTransactionOperation Operation, RotationOutcomeFence Fence,
            RegistryCommitReceipt Receipt, VerifiedRegistryHead Head, RotationRecoveryAuthority Authority) : RotationRecoveryEvidence(Authority);
        // Trusted completed original-binding prior restoration serialized with the fence
        // and authoritative parent. Matching visuals cannot supersede a possibly committed
        // target: the verifier must establish the entire serialized outcome.
        public sealed record PriorRestoredAndFenced(RotationTransactionOperation Operation, RotationOutcomeFence Fence,
            HeroVerifiedVisual Restoration, RegistryCommitReceipt Receipt, VerifiedRegistryHead Head,
            RotationRecoveryAuthority Authority) : RotationRecoveryEvidence(Authority);
        public sealed record VisualClosure(RotationTransactionOperation Operation, TransactionEvent.CompletionCommitted Completion,
            RotationRecoveryAuthority Authority) : RotationRecoveryEvidence(Authority);
        // All old deliveries fenced, Apply never ran (Prepare/ARM may have occurred).
        // A head observation alone is insufficient.
        public sealed record VisualFencedNotExecuted(RotationTransactionOperation Operation, VerifiedRegistryHead Head,
            RotationRecoveryAuthority Authority, RegistryCommitReceipt Receipt = null) : RotationRecoveryEvidence(Authority);
        public sealed record ModeClosure(ModeOperation Operation, RotationModeEvent.ModeCommitted Completion,
            RotationRecoveryAuthority Authority) : RotationRecoveryEvidence(Authority);
        // Exact intent never consumed/executed; all its execution deliveries fenced.
        public sealed record ReadinessFencedNotExecuted(RotationReadyIntent Intent, VerifiedRegistryHead Head,
            RotationRecoveryAuthority Authority) : RotationRecoveryEvidence(Authority);
    }
    // One bounded snapshot: Before.LastRecovery must be null. No resource release authority.
    public sealed record RotationRecoveryRecord(RotationTransactionState Before, RotationRecoveryEvidence Evidence);
    public abstract record RotationTransactionEvent
    {
        private RotationTransactionEvent() { }
        public sealed record Recover(RotationRecoveryEvidence Evidence) : RotationTransactionEvent;
        public sealed record Mode(RotationModeEvent Event) : RotationTransactionEvent;
        public sealed record ConsumeReady(RotationReadyIntent Intent) : RotationTransactionEvent;
        public sealed record Transaction(HeroBindingToken Hero, TransactionEvent Event) : RotationTransactionEvent;
    }
    public abstract record RotationTransactionCommand
    {
        private RotationTransactionCommand() { }
        public sealed record Mode(RotationModeCommand Command) : RotationTransactionCommand;
        // Caller checks original hero AND the inner command's skin correlation.
        public sealed record Transaction(HeroBindingToken Hero, SkinCommand Command) : RotationTransactionCommand;
    }
    public sealed class RotationTransactionDecision
    {
        public RotationTransactionState State { get; }
        public string Diagnosis { get; }
        public IReadOnlyList<RotationTransactionCommand> Commands { get; }
        public RotationTransactionDecision(RotationTransactionState state, string diagnosis,
            IEnumerable<RotationTransactionCommand> commands = null)
        { State=state; Diagnosis=diagnosis; Commands=Array.AsReadOnly(commands?.ToArray() ?? Array.Empty<RotationTransactionCommand>()); }
    }
}
